src/client/shader.cpp

```diff
@@ -280,6 +285,50 @@ class MainShaderConstantSetter : public IShaderConstantSetter
 		};
 		m_normal.set(m, services);
 #endif
+
+		// Set Shadow shader uniform
+		ShadowRenderer *shadow = m_shadow_map_enabled ?  RenderingEngine::get_shadow_renderer() : nullptr;
+		if (shadow) {
+			core::matrix4 shadowViewProj = shadow->getDirectionalLight()
+				.getProjectionMatrix();
+			shadowViewProj *= shadow->getDirectionalLight().getViewMatrix();
+
+			services->setPixelShaderConstant(
+				services->getPixelShaderConstantID("m_ShadowViewProj"),
+				*reinterpret_cast<float(*)[16]>(shadowViewProj.pointer()), 16);
+
+			float v_LightDirection[3];
+			shadow->getDirectionalLight().getDirection().getAs3Values(v_LightDirection);
+			services->setPixelShaderConstant(
+					services->getPixelShaderConstantID("v_LightDirection"),
+					*reinterpret_cast<float(*)[3]>(v_LightDirection), 3);
+
+			float TextureResolution = (float)shadow->getDirectionalLight()
+				.getMapResolution();
+			services->setPixelShaderConstant(
+					services->getPixelShaderConstantID("f_textureresolution"),
+					&TextureResolution, 1);
+
+			float ShadowStrengh = (float)shadow->getShadowStrengh();
+			services->setPixelShaderConstant(
+					services->getPixelShaderConstantID("f_shadow_strength"),
+					&ShadowStrengh, 1);
+
+			float timeofDay = (float)shadow->getTimeofDay();
+			services->setPixelShaderConstant(
+					services->getPixelShaderConstantID("f_timeofday"),
+					&timeofDay, 1);
+
+			float shadowFar = shadow->getMaxShadowFar();
+			services->setPixelShaderConstant(
+					services->getPixelShaderConstantID("f_shadowfar"),
+					&shadowFar, 1);
```

Comment:
these should use `CachedVertexShaderSetting` like the others

Reply:
IDK. I guess we have to change them, yes. But for now I think it's quite convenient to use it because adding/removing uniforms to test is much faster and easy in this way.